OIT step 1: Refactor renderer

almarklein opened the pull request:
This refactors the wgpu render code to prepare for implementing OIT, as planned in #196.

Todo:
* [x] Refactor so all blending stuff is defined in a single place.
* [x] More refactoring to support 1 or 2 render passes. 
* [x] Implement render mode that we had before this.
* [x] Implement simple 2-pass render mode.
* [x] Add `renderer.blend_mode`.
* [x] Decide how to implement the two fragment entry points with the shader (templates). Depends on https://github.com/gfx-rs/naga/issues/1509 -> see #200 
* [x] Implement for all materials.
* [x] Fix example to use PySide6.
* [x] Remove the `with open("c:/dev/py/tmp.wgsl")` thing.
* [x] Self-review + check for todo's.
* [x] Relief limitation that varying assignment must be a oneliner
* [x] Cleanup

berendkleinhaneveld commented:
Well, you could at least call it `FragmentBlender` :)

berendkleinhaneveld commented on pygfx/renderers/wgpu/_pipelinebuilder.py:
```diff
@@ -0,0 +1,520 @@
+            if pipeline_info.get("target", None) is None:
+                render_pipelines.append(pipeline)
+            else:
+                raise NotImplementedError("Alternative render pipelines")
```
This should not be here, right?

almarklein replied:
The idea is that material renderers can chose to render to an alternative render target (multi-pass rendering). We don't need it (yet) so it has not been implemented.

almarklein replied:
This code was already there, BTW. It just moved.

berendkleinhaneveld replied:
Thanks for clarifying!

almarklein commented:
Still refactoring and cleaning up, but here are some results for three initial blend modes.

The "solid" blend mode pretends that there is no such thing as transparency. Only included as a stub. It has one benefit: it's order independent :P

![image](https://user-images.githubusercontent.com/3015475/140046828-a33cd757-855e-41f1-b0a7-32a3162dee3a.png)

The "minimal" blend mode uses one pass, processing opaque and transparent fragments at the same time, while depth-testing. This is basically what we had. You can see how the red plane (the first-drawn) hides the other planes behind it because of the depth test. Also the blue plane (drawn last) disappears behind the green plane. Oh, and the sphere is partially hidden too.

![image](https://user-images.githubusercontent.com/3015475/140048522-71b16155-cb55-44f8-b11f-001ce2143737.png)

The "simple" blend mode uses two passes. One for opaque and one for transparent fragments. The looks result is much better. However, it's not order-independent, as can be seen in the second image.

![image](https://user-images.githubusercontent.com/3015475/140048199-0195748c-3ab4-47e5-ad2e-c1f080742ea6.png)
![transparency01](https://user-images.githubusercontent.com/3015475/140049158-cecb6564-145e-410c-80f0-8751d5767639.gif)
)

Korijn commented:
Just wondering if we are still sorting objects along their depth in camera space prior to rendering? That way at least the result will be deterministic, though still order-dependent.

almarklein commented:
I think we are.

almarklein commented:
I'm now happy with the approach to deal with the shaders that run (slightly changed) in multiple render passes. See #200 for details. Other than that, we just create two wgsl samples, and two wgpu shader modules: one for each render pass. I initially wanted to include all entrypoints into the same shader, but that quickly becomes complicated to implement and explain.

almarklein commented:
This is ready from my end.

Some updates:
* I implemented the new approach for all material-renderers! Ran into a few hurdles though:
* Since the blend mode affects the pipeline of the wobjects, the pipeline data could no longer be stored on the wobjects themselves (it would go wrong if using one object in multiple renders with different blend modes). Therefore the pipeline-data is now associated via weak-ref dicts. For details see the extensive comment in _renderer.py.
* The blender defines the `FragmentOutput`, but it does not know beforehand whether the eventual shader wants to also write depth. I added another (much simpler) wgsl post processing pass to resolve this.
* The varyings resolve post-processing now deals with multiline assignments.

Overall, this refactoring is mainly a preparation for more advanced blend modes, adding support for multi-stage rendering. Because we now have a 2 pass blend mode, even though its still simple/wrong, it already yields much better results for scenes that contain transparent objects. It also deals much better with e.g. the semitransparent edges of our lines (for aa); our lines look much better now!

Korijn commented:
> * Since the blend mode affects the pipeline of the wobjects, the pipeline data could no longer be stored on the wobjects themselves (it would go wrong if using one object in multiple renders with different blend modes). Therefore the pipeline-data is now associated via weak-ref dicts. For details see the extensive comment in _renderer.py.

Sounds like a big improvement to me!

> * The blender defines the `FragmentOutput`, but it does not know beforehand whether the eventual shader wants to also write depth. I added another (much simpler) wgsl post processing pass to resolve this.

How should I feel about this? :)

> * The varyings resolve post-processing now deals with multiline assignments.

Awesome!

> Overall, this refactoring is mainly a preparation for more advanced blend modes, adding support for multi-stage rendering. Because we now have a 2 pass blend mode, even though its still simple/wrong, it already yields much better results for scenes that contain transparent objects. It also deals much better with e.g. the semitransparent edges of our lines (for aa); our lines look much better now!

Let's merge and continue in another PR!

almarklein commented:
> How should I feel about this? :)

It means we can write:
```
var out = finalize_fragment();
out.depth = ndc_pos.z / ndc_pos.w;
```
and it Just Works.

We could make it more explicit, e.g.:
```
var out = finalize_fragment_at_depth(ndc_pos.z / ndc_pos.w;);
```

BTW, I think I may replace `add_fragment` and `finalize_fragment` with a single function. Let's revisit then. I'll add a note in the TODO of #196

Korijn commented:
As far as I am concerned we can merge :)

almarklein commented:
Wasn't sure whether @berendkleinhaneveld wanted to have a look.

berendkleinhaneveld commented:
Please go ahead and merge: I'll catch up later!
